### PR TITLE
Milestone 5

### PR DESCRIPTION
**Pull request by frederico-miranda**
This is a pull request opened for receiving a review for the fifth and final milestone of the 'React & Redux - Calculator' project. This fifth milestone consists of implementing the event listeners that make the app do its thing and behave like a real calculator!

The instructions for this milestone can be found at the following link:
[React & Redux - Calculator - Milestone 5](https://github.com/microverseinc/project-react-calculator/blob/master/milestones/MILESTONE_5.md)

The link for the deployed Heroku application is the following one:
[Heroku App - React & Redux - Calculator](https://polar-shelf-49767.herokuapp.com/)

You can also test this on your local machine by cloning this repository, installing the npm modules and starting the server. The following list of commands may help you to do so:

```
git clone https://github.com/frederico-miranda/react-calculator/
cd react-calculator
git checkout milestone-5
npm install
npm start
```

NOTE 1: The specification instructs me to use an event listener for each button, but I consciously decided to not do that because there is a better pattern for this use case. There are use cases where there are hundreds of DOM elements (a large grid with clickable rectangles, as an example), and by using the bubbling of DOM events, it's possible to have **one** parent with **one** event listener handle all the clicks on the elements (instead of using an event listener for each rectangle, as is the pattern given by the instructions). When you look at the code, you'll see this pattern instead of the pattern of the specification.

NOTE 2: The App state should have been passed to the ButtonPanel, but I did not do that. I have no good reason for justifying this other than my lack of experience with React. Now that I think about it, using a context object might have helped here, but I'm still learning React.

### CODE REVIEW

File: src/components/button.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ const Button = ({ name, wide, color }) => {
     flex: wide ? '2 0 1px' : '1 0 1px',
     background: color,
   };
-  return (<button type="button" style={inlineStyle}>{name}</button>);
+  return (<button type="button" data-name={name} style={inlineStyle}>{name}</button>);
```
Comment (bolah2009) [inline src/components/button.js]: Passing down `clickHandler` as a prop from the `ButtonPanel` component, we can have:

```jsx
(<button type="button" onClick={clickHandler} style={inlineStyle}>{name}</button>);
```
#### Requirement Screenshot

image


With these, the `ButtonPanel` should have a function that would be passed down to the `Button` component. The function will take the `buttonName` as an argument:

```diff
- const ButtonPanel = ({ className }) => (
+ const ButtonPanel = ({ clickHandler, className }) => (
  <div className={className}>
    {
      buttonRows.map((row, indexA) => {
        const rowKey = `BUTTON_PANEL[${indexA}]`;
        return (
          <div key={rowKey}>
            {
              row.map(({ name, color, wide }, indexB) => {
                const buttonKey = `${rowKey}[${indexB}]`;
-               return (<Button key={buttonKey} name={name} color={color} wide={wide} />);
+               return (<Button clickHandler={() => clickHandler(name)} key={buttonKey} name={name} color={color} wide={wide} />);
              })
            }
          </div>
        );
      })
    }
  </div>
);
```


#### Requirement Screenshot

image

File: src/components/app.js [hunk 2]
```diff
@@ -1,15 +1,70 @@
...
+  };
+
+  let result;
+  if (next !== null) {
+    result = `${total.toString()}${operation}${next.toString()}`;
+  } else if (operation !== null) {
+    result = `${total.toString()}${operation}`;
+  } else if (total !== null) {
+    result = `${total.toString()}`;
+  }
```
[the remark this answers is not in the document]
Comment (frederico-miranda) [reply src/components/app.js]: This has been addressed. 'calculate.js' now has a lambda called 'displayString' which does this formatting of an object into a string.

File: src/components/button.js (hunk 1 above)
Comment (frederico-miranda) [reply src/components/button.js]: This has been addressed.

File: src/components/app.js [hunk 3]
```diff
@@ -1,15 +1,70 @@
...
+    /* This elements onClick handler is present for catching the events that    */
+    /* bubble up from its 15 child buttons.                                     */
+    /* Having *one* event listener is better than having many, specially so     */
+    /* when the children are uniform (can share the same event listener).       */
+    /* eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions */
+    <div role="group" onClick={handleClick} onKeyUp={handleKeyUp} className={className}>
+      <Display className="calculator-display" result={result} />
+      <ButtonPanel className="calculator-button-panel" />
+    </div>
+  );
```
[the remark this answers is not in the document]
Comment (frederico-miranda) [reply src/components/app.js]: This has been addressed.

**Comment by frederico-miranda** [comment]
I believe I have addressed the issues you have raised, so I'm requesting another code review.

**Pull request merged**